File: 46. Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        List<Integer> myArray = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        find(nums, output, myArray, used);
        return output;
        
    }
    
    public void find(int[] nums, List<List<Integer>> output, List<Integer> input, boolean[] used){
        if (input.size() == nums.length) {
            output.add(new ArrayList<>(input));
        } else {
            for (int i = 0; i < nums.length; i++) {
                if (used[i] == false) {
                    input.add(nums[i]);
                    used[i] = true;
                    find(nums, output, input, used);
                    input.remove(input.size() - 1);
                    used[i] = false;
                }
            }
        }
    }
}

File: 46. Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        output = []
        
        for num in nums:
            self.find_permute(nums, [num], set([num]), output)
        return output
    
    def find_permute(self, nums, arr, visited, permutations):
        if len(arr) == len(nums):
            permutation = arr[:]
            permutations.append(permutation)
        else:
            for num in nums:
                if num not in visited:
                    arr.append(num)
                    visited.add(num)
                    self.find_permute(nums, arr, visited, permutations)
                    visited.remove(num)
                    arr.pop()
